src/Microsoft.ComponentDetection.Detectors/rust/RustCliDetector.cs

```diff
@@ -0,0 +1,134 @@
+namespace Microsoft.ComponentDetection.Detectors.Rust;
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.ComponentDetection.Contracts;
+using Microsoft.ComponentDetection.Contracts.Internal;
+using Microsoft.ComponentDetection.Contracts.TypedComponent;
+using Microsoft.ComponentDetection.Detectors.Rust.Contracts;
+using Microsoft.Extensions.Logging;
+
+/// <summary>
+/// A Rust CLI detector that uses the cargo metadata command to detect Rust components.
+/// </summary>
+public class RustCliDetector : FileComponentDetector, IExperimentalDetector
+{
+    private readonly ICommandLineInvocationService cliService;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="RustCliDetector"/> class.
+    /// </summary>
+    /// <param name="componentStreamEnumerableFactory">The component stream enumerable factory.</param>
+    /// <param name="walkerFactory">The walker factory.</param>
+    /// <param name="cliService">The command line invocation service.</param>
+    /// <param name="logger">The logger.</param>
+    public RustCliDetector(
+        IComponentStreamEnumerableFactory componentStreamEnumerableFactory,
+        IObservableDirectoryWalkerFactory walkerFactory,
+        ICommandLineInvocationService cliService,
+        ILogger<RustCliDetector> logger)
+    {
+        this.ComponentStreamEnumerableFactory = componentStreamEnumerableFactory;
+        this.Scanner = walkerFactory;
+        this.cliService = cliService;
+        this.Logger = logger;
+    }
+
+    /// <inheritdoc />
+    public override string Id => "RustCli";
+
+    /// <inheritdoc />
+    public override IEnumerable<string> Categories { get; } = new[] { "Rust" };
+
+    /// <inheritdoc />
+    public override IEnumerable<ComponentType> SupportedComponentTypes => new[] { ComponentType.Cargo };
+
+    /// <inheritdoc />
+    public override int Version => 1;
+
+    /// <inheritdoc />
+    public override IList<string> SearchPatterns { get; } = new[] { "Cargo.toml" };
+
+    /// <inheritdoc />
+    protected override async Task OnFileFoundAsync(ProcessRequest processRequest, IDictionary<string, string> detectorArgs)
+    {
+        var componentStream = processRequest.ComponentStream;
+
+        try
+        {
+            if (!await this.cliService.CanCommandBeLocatedAsync("cargo", null))
+            {
+                this.Logger.LogWarning("Could not locate cargo command. Skipping Rust CLI detection");
+                return;
+            }
+
+            var cliResult = await this.cliService.ExecuteCommandAsync(
+                "cargo",
+                null,
+                "metadata",
+                "--manifest-path",
+                componentStream.Location,
+                "--format-version=1",
+                "--locked");
```

Comment:
Prevents us from updating the `Cargo.lock`: https://doc.rust-lang.org/cargo/commands/cargo-install.html#dealing-with-the-lockfile

Reply:
I suppose there is a tradeoff here that the command will fail if the lockfile needs to be updated. 

Our similar use of `-mod=readonly` in the go detector causes failures fairly often in the go detector.

Reply:
We should probably add a log here printing stderr here if we fail then.

Reply:
Agreed

Reply:
From https://doc.rust-lang.org/cargo/commands/cargo-metadata.html#manifest-options:
>Either of these flags requires that the Cargo.lock file is up-to-date. If the lock file is missing, or it needs to be updated, Cargo will exit with an error. The --frozen flag also prevents Cargo from attempting to access the network to determine if it is out-of-date.